uniffi/src/support/tests.rs

```diff
@@ -0,0 +1,135 @@
+    if let Some(cdylib_file) = compiled_components.get(pkg_dir) {
+        return Ok(cdylib_file.to_string());
+    }
+    // Nope, looks like we'll have to compile it afresh.
+    let mut cmd = Command::new("cargo");
+    cmd.arg("build").arg("--message-format=json").arg("--lib");
+    cmd.current_dir(pkg_dir);
+    cmd.stdout(Stdio::piped());
+    let mut child = cmd.spawn()?;
+    let output = std::io::BufReader::new(child.stdout.take().unwrap());
+    // Build the crate, looking for any cdylibs that it might produce.
+    let cdylibs = cargo_metadata::Message::parse_stream(output)
+        .filter_map(|message| match message {
+            Err(e) => Some(Err(e.into())),
+            Ok(Message::CompilerArtifact(artifact)) => {
+                if artifact.target.kind.iter().any(|item| item == "cdylib") {
+                    Some(Ok(artifact))
+                } else {
+                    None
+                }
+            }
+            _ => None,
+        })
+        .collect::<Result<Vec<_>>>()?;
+    if !child.wait()?.success() {
+        bail!("Failed to execute `cargo build`");
+    }
+    // If we didn't just build exactly one cdylib, we're going to have a bad time.
+    match cdylibs.len() {
+        0 => bail!("Crate did not produce any cdylibs, it must not be a uniffi component"),
+        1 => (),
+        _ => bail!("Crate produced multiple cdylibs, it must not be a uniffi component"),
+    }
+    if cdylibs[0].filenames.len() != 1 {
+        bail!(
+            "Crate did not produce exactly one cdylib filename, it must not be a uniffi component"
+        );
+    }
+    let cdylib_file = cdylibs[0].filenames[0].to_string_lossy().into_owned();
+    // Cache the result for subsequent tests.
+    compiled_components.insert(pkg_dir.to_string(), cdylib_file.clone());
+    Ok(cdylib_file)
+}
+
+/// Ensure that a uniffi component cdylib has foreign language bindings generated and ready for use.
+///
+/// This function takes the path to a uniffi component cdylib, a directory into which foreign language
+/// bindings should be generated, and select target language. It ensures that the cdylib has bindings
+/// for that language generated, compiled, and ready for use.
+///
+/// Internally, this function does a bit of caching and concurrency management to avoid rebuilding
+/// the component for multiple testcases.
+pub fn ensure_generated_bindings<P1, P2, L>(cdylib_file: P1, out_dir: P2, language: L) -> Result<()>
+where
+    P1: AsRef<Path>,
+    P2: AsRef<Path>,
+    L: TryInto<bindings::TargetLanguage, Error = anyhow::Error>,
+{
+    let cdylib_file: PathBuf = cdylib_file.as_ref().to_path_buf();
+    let out_dir: PathBuf = out_dir.as_ref().to_path_buf();
+    let language: bindings::TargetLanguage = language.try_into()?;
+    // Have we already generated these bindings?
+    let mut generated_bindings = GENERATED_BINDINGS.lock().unwrap();
+    // XXX TODO: figure out correct type of `HashSet` so that I don't need to `clone()` here...
```

Comment:
Ooh, good question! I wonder if this will need a tuple struct with a custom `Borrow<&(&PathBuf, &PathBuf, &TargetLanguage)>` implementation, or if you can sprinkle the right combination of `&`s  that’ll make the compiler understand this...